### PR TITLE
Shaders using GLSL not Cg/HLSL

### PR DESCRIPTION
**Pull request by revmischa**
Trying to get rid of super ancient unsupported nVidia Cg framework and use standard GL shaders.
The shaders in preset files are written in HLSL for DirectX. projectM used a toolkit from nVidia called Cg to handle them, but Cg is long deprecated and obsolete.
I'm attempting to use https://github.com/Thekla/hlslparser to transpile from HLSL to GLSL on the fly, and it's close to working but has some syntax issues.

The main conversion is done in:
https://github.com/projectM-visualizer/projectm/pull/26/files#diff-6df95f02d59c83baf42ed81c8668722a
https://github.com/projectM-visualizer/projectm/pull/26/files#diff-f4b5c1904e3d8fa6c1248fe1ac843d79

This may be close to working but it has issues right now. Needs work.

### CODE REVIEW

**Comment by axe312ger** [comment]
Awesome! Will this boost the visual quality, too? Or is it just refactoring to get rid of legacy code?

**Comment by revmischa** [comment]
If we can get the shaders to work it will greatly increase the quality. However the shaders are written in DirectX/Cg style and may need some converting to GLSL style. I dunno anything about doing that so someone else needs to help.

**Comment by axe312ger** [comment]
Maybe twitter can help? I tweeted about it here: https://twitter.com/axe312ger/status/968429518095634432

**Comment by revmischa** [comment]
I'll put up a status update on where things are at in the next few weeks.. stay tuned

**Comment by revmischa** [comment]
I'm not working on anything right now for projectM (busy with `$dayjob`), but I'm happy to provide guidance to anyone who wants to contribute

**Comment by axe312ger** [comment]
Any chance this can help?
https://www.opengl.org/discussion_boards/showthread.php/174728-Compiling-CG-to-GLSL

`cgc -profile glslv vert.cg`

**Comment by revmischa** [comment]
Very cool! Feel free to give it a try. The shader code is shoved inside the preset files in a funky format though so you'd have to parse it out, convert it, and reinsert it

**Comment by revmischa** [comment]
I'm giving https://github.com/Thekla/hlslparser a try - it should transpile from HLSL to GLSL. I've got it hooked up, now work needs to be done on passing in the variables and texture samplers from projectM into the shaders. Then maybe shaders will work!

**Comment by revmischa** [comment]
Some syntax issues and GL versioning issues but it's close! I think!

**Comment by axe312ger** [comment]
I have no idea what you are doing here but 👍and keep the 🚀flying 🙈

File: src/libprojectM/Renderer/ShaderEngine.cpp [hunk 1]
```diff
@@ -122,10 +122,63 @@ ShaderEngine::ShaderEngine()
     UNIFORM_V2F_C4F_VERTEX_POINT_SIZE = glGetUniformLocation(programID_v2f_c4f, "vertex_point_size");
     UNIFORM_V2F_C4F_T2F_VERTEX_TRANFORMATION = glGetUniformLocation(programID_v2f_c4f_t2f, "vertex_transformation");
     UNIFORM_V2F_C4F_T2F_FRAG_TEXTURE_SAMPLER = glGetUniformLocation(programID_v2f_c4f_t2f, "texture_sampler");
+
+
+
+	std::string line;
+	std::ifstream myfile(DATADIR_PATH "/shaders/projectM.cg");
```
Comment (revmischa) [inline src/libprojectM/Renderer/ShaderEngine.cpp]: can we try not reading the shaders from files? it simplifies packaging. just stick them in header files if possible.

**Comment by revmischa** [comment]
@deltaoscarmike what's next?

**Comment by axe312ger** [comment]
😍 can I send you guys beer or sth? 😁

File: src/libprojectM/Renderer/ShaderEngine.cpp [hunk 2]
```diff
@@ -805,237 +749,317 @@ void ShaderEngine::SetupShaderVariables(GLuint program, const Pipeline &pipeline
...
-			{
-			{ 0, 0.25 },
-			{ 0, 0 },
-			{ 0.25, 0 },
-			{ 0.25, 0.25 } };
 
-			glVertexPointer(2, GL_FLOAT, 0, points);
-			glBlendFunc(GL_ONE,GL_ZERO);
-			glDrawArrays(GL_TRIANGLE_FAN, 0, 4);
+void ShaderEngine::RenderBlurTextures(const Pipeline &pipeline, const PipelineContext &pipelineContext)
```
Comment (revmischa) [inline src/libprojectM/Renderer/ShaderEngine.cpp]: maybe the blur stuff should go in a separate file. this file is getting waaayyy too big

File: src/libprojectM/Renderer/Texture.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    for(std::vector<Sampler*>::const_iterator iter = samplers.begin(); iter != samplers.end(); iter++)
+    {
+        if ((*iter)->wrap_mode == _wrap_mode && (*iter)->filter_mode == _filter_mode)
+        {
+            return *iter;
+        }
+    }
+
+    // Sampler not found -> adding it
+    Sampler * sampler = new Sampler(_wrap_mode, _filter_mode);
```
Comment (revmischa) [inline src/libprojectM/Renderer/Texture.cpp]: does a text ever have more than one sampler?

File: src/projectM-sdl/pmSDL.cpp [hunk 4]
```diff
@@ -228,15 +236,107 @@ projectMSDL::projectMSDL(std::string config_file, int flags) : projectM(config_f
...
-void projectMSDL::init(SDL_Window *window, SDL_GLContext *_glCtx) {
+void projectMSDL::init(SDL_Window *window, SDL_GLContext *_glCtx, const bool _renderToTexture) {
     win = window;
     glCtx = _glCtx;
-    selectRandom(true);
     projectM_resetGL(width, height);
+
+    renderToTexture = _renderToTexture;
+
+    if (renderToTexture) {
```
Comment (revmischa) [inline src/projectM-sdl/pmSDL.cpp]: this code shouldn't be in pm-sdl right? should be in the library?

File: src/projectM-sdl/pmSDL.cpp [hunk 5]
```diff
@@ -228,15 +236,107 @@ projectMSDL::projectMSDL(std::string config_file, int flags) : projectM(config_f
...
-    return std::string("hey");
+    unsigned int index = 0;
+    if (selectedPresetIndex(index)) {
+        return getPresetName(index);
+    }
+    return std::string();
+}
+
+
+void projectMSDL::renderTexture() {
```
Comment (revmischa) [inline src/projectM-sdl/pmSDL.cpp]: i don't see anything sdl-related here

File: src/libprojectM/Renderer/TextureManager.cpp [hunk 6]
```diff
@@ -23,253 +22,375 @@
...
+    }
 
- Preload();
- loadTextureDir();
+    PerlinNoise noise;
+
+    GLuint noise_texture_lq_lite;
+    glGenTextures(1, &noise_texture_lq_lite);
+    glBindTexture(GL_TEXTURE_2D, noise_texture_lq_lite);
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, 32, 32, 0, GL_LUMINANCE, GL_FLOAT, noise.noise_lq_lite);
```
Comment (revmischa) [inline src/libprojectM/Renderer/TextureManager.cpp]: Not defined for me

screen shot 2018-07-23 at 12 58 27

File: src/libprojectM/Renderer/hlslparser/src/HLSLTree.h [hunk 7]
```diff
@@ -513,6 +517,22 @@ struct HLSLArgument : public HLSLNode
...
+    static const HLSLNodeType s_type = HLSLNodeType_Macro;
+    HLSLMacro()
+    {
+        name            = NULL;
+        argument        = NULL;
+        macroAliased    = NULL;
+    }
+    const char*         name;
+    HLSLArgument*       argument;
+    std::string         value;
```
Comment (revmischa) [inline src/libprojectM/Renderer/hlslparser/src/HLSLTree.h]: this errors for me

screen shot 2018-07-23 at 13 20 35

Comment (revmischa) [reply src/libprojectM/Renderer/hlslparser/src/HLSLTree.h]: easily fixed by #including  `<string>`

**Comment by revmischa** [comment]
Has a weird blue bar at the bottom (and sometimes top now)

screen shot 2018-07-23 at 13 28 54

**Comment by revmischa** [comment]
```
presets/An AdamFX n Martin Infusion 2 flexi - Why The Sky Looks Diffrent Today - AdamFx n Martin Infusion - Tack Tile Disfunction B.milk(85) : Syntax error: expected ';' near '%2'
Failed to parse HLSL (Warp) shader
```

```
stahlregen + geiss + shifter - babylon.milk(85) : Syntax error: expected ')' near ','
Failed to parse HLSL (Comp) shader
```

```
Texture loading error for: worms
Failed to compile shader 'Fragment: Warp'. Error: ERROR: 0:90: Samplers not allowed except in non-buffer uniforms and in-parameters
ERROR: 0:107: Left-hand-side of assignment must not be read-only
```

**Comment by revmischa** [comment]
This looks freakin amazing and excellent.

Some things render wrong, or add a big weird border. Here's some things I saw:

screen shot 2018-07-23 at 19 28 13
screen shot 2018-07-23 at 19 28 20
screen shot 2018-07-23 at 19 35 12



But otherwise it looks awesomee

**Comment by revmischa** [comment]
screen shot 2018-07-23 at 19 28 29

border and weird shader (maybe because I changed GL_LUMINENCE to 0..)

File: src/libprojectM/Renderer/Texture.cpp (hunk 3 above)
Comment (deltaoscarmike) [reply src/libprojectM/Renderer/Texture.cpp]: Main texture and user textures have 4 samplers, see [texture sampling](http://www.geisswerks.com/milkdrop/milkdrop_preset_authoring.html#3f8)

File: src/projectM-sdl/pmSDL.cpp (hunk 4 above)
Comment (deltaoscarmike) [reply src/projectM-sdl/pmSDL.cpp]: I don't think so, it is left to the developer to decide where and how he will draw the projectM generated texture.

File: src/projectM-sdl/pmSDL.cpp (hunk 5 above)
Comment (deltaoscarmike) [reply src/projectM-sdl/pmSDL.cpp]: It's a render to texture example/POC.

**Comment by revmischa** [comment]
This is coming along great, thanks so much @deltaoscarmike. There are a number of visual issues I saw that I posted above, and a decent number of shaders failing to compile. If we can get those addressed then we should merge and release this soon. It's a massive improvement. I think the biggest problem I keep seeing is a big chunk of blue that takes up some amount of the bottom half of the screen from time to time. Sometimes just a sliver at the bottom and sometimes half the screen is this blue color. Maybe related to the weird borders

**Comment by revmischa** [comment]
Testing it out with emscripten with `--enable-gles`...
```
INFO: GL_VERSION: OpenGL ES 2.0 (WebGL 1.0 (2.1 ATI-1.68.20))
INFO: GL_SHADING_LANGUAGE_VERSION: OpenGL ES GLSL ES 1.00 (WebGL GLSL ES 1.0 (1.20))
INFO: GL_VENDOR: WebKit
INFO: Config file not found, using development settings
Failed to compile shader 'Vertex: v2f_c4f'. Error: ERROR: unsupported shader version

Failed to compile shader 'Fragment: v2f_c4f'. Error: ERROR: unsupported shader version

Failed to link program: 
Validation Failed: Link error:
  Program is not successfully linked.

Failed to compile shader 'Vertex: v2f_c4f_t2f'. Error: ERROR: unsupported shader version

Failed to compile shader 'Fragment: v2f_c4f_t2f'. Error: ERROR: unsupported shader version

Failed to link program: 
Validation Failed: Link error:
  Program is not successfully linked.

Failed to compile shader 'Vertex: blur1'. Error: ERROR: unsupported shader version

Failed to compile shader 'Fragment: blur1'. Error: ERROR: 0:39: 'assign' : cannot convert from 'const int' to 'mediump float'

Failed to link program: 
Validation Failed: Link error:
  Program is not successfully linked.

Failed to compile shader 'Vertex: blur2'. Error: ERROR: unsupported shader version

Failed to compile shader 'Fragment: blur2'. Error: ERROR: 0:30: '-' : wrong operand types - no operation '-' exists that takes a left-hand operand of type 'const int' and a right operand of type 'mediump float' (or there is no acceptable conversion)
ERROR: 0:30: 'min' : no matching overloaded function found
ERROR: 0:35: 'assign' : cannot convert from 'const int' to 'mediump float'

Failed to link program: 
Validation Failed: Link error:
  Program is not successfully linked.
```

**Comment by revmischa** [comment]
Also emscripten complains:
```
warning: unresolved symbol: glBindSampler
warning: unresolved symbol: glDeleteSamplers
warning: unresolved symbol: glGenSamplers
warning: unresolved symbol: glSamplerParameteri
warning: unresolved symbol: glTexImage3D
warning: unresolved symbol: glUniformMatrix4x3fv
```

File: src/projectM-emscripten/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+-DSYSCONFDIR=\""$(sysconfdir)"\" \
+-I${top_srcdir}/src/libprojectM \
+-I${top_srcdir}/src/libprojectM/Renderer \
+USE_SDL=2 \
+${SDL_CFLAGS}
+
+REALSRCDIR=${top_srcdir}/src/projectM-sdl
+
+bin_PROGRAMS = projectW.bc
+projectW_bc_SOURCES = $(REALSRCDIR)/pmSDL.cpp $(REALSRCDIR)/projectM_SDL_main.cpp $(REALSRCDIR)/pmSDL.hpp
```
Comment (deltaoscarmike) [inline src/projectM-emscripten/Makefile.am]: projectM_SDL_emscripten.cpp should be deleted ?

Comment (revmischa) [reply src/projectM-emscripten/Makefile.am]: Yeah most likely true. `projectM_SDL_emscripten.cpp` is from long long ago. Will need to look into this at some point.

**Comment by revmischa** [comment]
blog post update https://wordpress.com/post/spiegelmock.com/1213

**Comment by revmischa** [comment]
Oh also most of the native presets segfault now...

**Comment by axe312ger** [comment]
I played around with the parameters and did end up with this:

`emcc -s USE_SDL=2 -s USE_WEBGL2=1 -s ALLOW_MEMORY_GROWTH=1 projectW.bc -o projectW.html`

Does not throw any warnings in the console.


Running it in the browser, it complains about failing to compile a shader:

```
projectW.js:1692 Fetch finished loading: GET "http://localhost:5000/projectW.wasm".
projectW:1249 Calling stub instead of sigaction()
projectW:1249 Calling stub instead of sigaction()
projectW:1249 Calling stub instead of sigaction()
projectW:1249 INFO: GL_VERSION: OpenGL ES 3.0 (WebGL 2.0 (OpenGL ES 3.0 Chromium))
projectW:1249 INFO: GL_SHADING_LANGUAGE_VERSION: OpenGL ES GLSL ES 3.00 (WebGL GLSL ES 3.00 (OpenGL ES GLSL ES 3.0 Chromium))
projectW:1249 INFO: GL_VENDOR: WebKit
projectW:1249 INFO: Config file not found, using development settings
projectW:1249 
projectW:1249 
projectW:1249 Failed to compile shader 'Fragment: blur1'. Error: ERROR: 0:39: 'assign' : cannot convert from 'const int' to 'mediump float'
projectW:1249 
projectW:1249 Failed to link program: invalid shaders
projectW:1249 program not linked
projectW:1249 Failed to compile shader 'Fragment: blur2'. Error: ERROR: 0:30: '-' : wrong operand types - no operation '-' exists that takes a left-hand operand of type 'const int' and a right operand of type 'mediump float' (or there is no acceptable conversion)
projectW:1249 ERROR: 0:30: 'min' : no matching overloaded function found
projectW:1249 ERROR: 0:35: 'assign' : cannot convert from 'const int' to 'mediump float'
projectW:1249 
projectW:1249 Failed to link program: invalid shaders
projectW:1249 program not linked
projectW:1249 [PresetLoader] ENOENT error. The path "/presets" probably does not exist. "man open" for more info.
projectW:1237 [PresetFactory] url is idle://Geiss & Sperl - Feedback (projectM idle HDR mix).milk
projectW:1237 Preset composite shader active: 0, preset warp shader active: 0
projectW:1237 No Textures Loaded from /usr/local/share/projectM/presets
projectW:1237 Preset composite shader active: 0, preset warp shader active: 0
projectW:1237 No Textures Loaded from /usr/local/share/projectM/presets
projectW:1237 Preset composite shader active: 0, preset warp shader active: 0
projectW:1249 INFO: Using audio driver: emscripten
projectW:1249 INFO: Found audio capture device 0: System audio capture device
projectW:1249 INFO: Opened audio capture device 2: System audio capture device
projectW:1249 INFO: Sample rate: 512, frequency: 48000, channels: 2, format: 33056
projectW:1237 fps: 60
```

**Comment by deltaoscarmike** [comment]
Status:
- All presets with a warp and/or composite shader are successfully compiled in GLSL
- Nativepresets are working again
- screens "borders" are fixed

To do: fix rendering differences against winamp/milkdrop. Some presets are correctly rendered but other have visual issues.

File: src/libprojectM/Renderer/TextureManager.cpp [hunk 9]
```diff
@@ -40,10 +40,13 @@ TextureManager::TextureManager(const std::string _presetsURL, const int texsizeX
     extensions.push_back(".dib");
 
     Preload();
-    loadTextureDir();
+
+    loadTextureDir(std::string(DATADIR_PATH) + "/presets");
+    loadTextureDir(std::string(DATADIR_PATH) + "/textures");
```
Comment (revmischa) [inline src/libprojectM/Renderer/TextureManager.cpp]: does this exist?

**Comment by revmischa** [comment]
Bravo! Good stuff 👍 
You think you're gonna submit the hlslparser fixes back?

**Comment by revmischa** [comment]
Got a crash:
```
libc++abi.dylib: Pure virtual function called!
Process 76589 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff5ba5bb66 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff5ba5bb66 <+10>: jae    0x7fff5ba5bb70            ; <+20>
    0x7fff5ba5bb68 <+12>: movq   %rax, %rdi
    0x7fff5ba5bb6b <+15>: jmp    0x7fff5ba52ae9            ; cerror_nocancel
    0x7fff5ba5bb70 <+20>: retq   
Target 0: (projectMSDL) stopped.
```

**Comment by revmischa** [comment]
doesn't crash if i disable threading

**Comment by deltaoscarmike** [comment]
> You think you're gonna submit the hlslparser fixes back?

No, my modifications have many impacts on Metal and HLSL generators and it's a lot of work to make sure these will still generate valid code.

>Got a crash:

Could you provide a more explicit backtrace callstack (with files and line number) ?

**Comment by revmischa** [comment]
Anything left to do @deltaoscarmike? Is it good to go? Can I help with anything?

**Comment by revmischa** [comment]
@deltaoscarmike merge y/n?

**Comment by deltaoscarmike** [comment]
Yes

**Pull request merged**